File: src/main/java/onion/compiler/environment/ReflectionalClassTypeRef.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.environment;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import onion.compiler.*;
import onion.compiler.Modifier;

/**
 * @author Kota Mizushima
 * Date: 2006/1/10
 */
public class ReflectionalClassTypeRef extends IRT.AbstractClassTypeRef {
  private static final String CONSTRUCTOR_NAME = "<init>";
  private Class klass;
  private ClassTable table;
  private int modifier;
  private MultiTable<IRT.MethodRef> methods;
  private OrderedTable<IRT.FieldRef> fields;
  private List<IRT.ConstructorRef> constructors;
  private OnionTypeBridge bridge;
  
  public ReflectionalClassTypeRef(Class klass, ClassTable table) {
    this.klass = klass;
    this.table = table;
    this.bridge = new OnionTypeBridge(table);
    this.modifier = toOnionModifier(klass.getModifiers());
  }

  public boolean isInterface() {
    return (klass.getModifiers() & java.lang.reflect.Modifier.INTERFACE) != 0;
  }

  public int modifier() {
    return modifier;
  }
  
  public String name() {
    return klass.getName();
  }

  public IRT.ClassTypeRef superClass() {
    Class superKlass = klass.getSuperclass();
    if(superKlass == null) return table.rootClass();
    IRT.ClassTypeRef superClass = table.load(superKlass.getName());
    if(superClass == this) return null;
    return superClass;
  }

  public IRT.ClassTypeRef[] interfaces() {
    Class[] interfaces = klass.getInterfaces();
    IRT.ClassTypeRef[] interfaceSyms = new IRT.ClassTypeRef[interfaces.length];
    for(int i = 0; i < interfaces.length; i++){
      interfaceSyms[i] = table.load(interfaces[i].getName());
    }
    return interfaceSyms;
  }
  
  public IRT.MethodRef[] methods() {
    requireMethodTable();
    return methods.values().toArray(new IRT.MethodRef[0]);
  }

  public IRT.MethodRef[] methods(String name) {
    requireMethodTable();
    return methods.get(name).toArray(new IRT.MethodRef[0]);
  }

  private void requireMethodTable() {
    if(methods == null){
      methods = new MultiTable<IRT.MethodRef>();
      for(Method method:klass.getMethods()){
        if(!method.getName().equals(CONSTRUCTOR_NAME)){
          methods.add(translate(method));
        }
      }
    }
  }
  
  public IRT.FieldRef[] fields() {
    requireFieldTable();
    return fields.values().toArray(new IRT.FieldRef[0]);
  }

  public IRT.FieldRef field(String name) {
    requireFieldTable();
    return fields.get(name);
  }

  private void requireFieldTable() {
    if(fields == null){
      fields = new OrderedTable<IRT.FieldRef>();
      for(Field field:klass.getFields()) {
        fields.add(translate(field));
      }
    }
  }

  public IRT.ConstructorRef[] constructors() {
    if(constructors == null){
      constructors = new ArrayList<IRT.ConstructorRef>();
      for(Constructor method:klass.getConstructors()) {
        constructors.add(translate(method));
      }
    }
    return constructors.toArray(new IRT.ConstructorRef[0]);
  }

  private static int toOnionModifier(int src){
    int modifier = 0;
    modifier |= (isOn(src, java.lang.reflect.Modifier.PRIVATE) ? Modifier.PRIVATE() : modifier);
    modifier |= (isOn(src, java.lang.reflect.Modifier.PROTECTED) ? Modifier.PROTECTED() : modifier);
    modifier |= (isOn(src, java.lang.reflect.Modifier.PUBLIC) ? Modifier.PUBLIC() : modifier);
    modifier |= (isOn(src, java.lang.reflect.Modifier.STATIC) ? Modifier.STATIC() : modifier);
    modifier |= (isOn(src, java.lang.reflect.Modifier.SYNCHRONIZED) ? Modifier.SYNCHRONIZED() : modifier);
    modifier |= (isOn(src, java.lang.reflect.Modifier.ABSTRACT) ? Modifier.ABSTRACT() : modifier);
    modifier |= (isOn(src, java.lang.reflect.Modifier.FINAL) ? Modifier.FINAL() : modifier);
    return modifier;
  }
  
  private static boolean isOn(int modifier, int flag){
    return (modifier & flag) != 0;
  }

  private IRT.MethodRef translate(Method method){
    Class[] arguments = method.getParameterTypes();
    IRT.TypeRef[] argumentRefs = new IRT.TypeRef[arguments.length];
    for (int i = 0; i < arguments.length; i++) {
      argumentRefs[i] = bridge.toOnionType(arguments[i]);
    }
    IRT.TypeRef returnRef = bridge.toOnionType(method.getReturnType());
    return new ClassFileMethodRef(toOnionModifier(method.getModifiers()), this, method.getName(), argumentRefs, returnRef);
  }
  
  private IRT.FieldRef translate(Field field){
    return new ClassFileFieldRef(toOnionModifier(field.getModifiers()),  this, field.getName(), bridge.toOnionType(field.getType()));
  }
  
  private IRT.ConstructorRef translate(Constructor constructor){
    Class[] arguments = constructor.getParameterTypes();
    IRT.TypeRef[] argumentRefs = new IRT.TypeRef[arguments.length];
    for (int i = 0; i < arguments.length; i++) {
      argumentRefs[i] = bridge.toOnionType(arguments[i]);
    }
    return new ClassFileConstructorRef(toOnionModifier(constructor.getModifiers()), this, "<init>", argumentRefs);
  }
}

File: src/main/scala/onion/compiler/environment/ReflectionalClassTypeRef.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005-2012, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.environment

import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.util.ArrayList
import java.util.List
import onion.compiler._
import onion.compiler.Modifier

/**
 * @author Kota Mizushima
 *
 */
object ReflectionalClassTypeRef {
  private def toOnionModifier(src: Int): Int = {
    ( (if (isOn(src, java.lang.reflect.Modifier.PRIVATE)) Modifier.PRIVATE else 0)
    | (if (isOn(src, java.lang.reflect.Modifier.PROTECTED)) Modifier.PROTECTED else 0)
    | (if (isOn(src, java.lang.reflect.Modifier.PUBLIC)) Modifier.PUBLIC else 0)
    | (if (isOn(src, java.lang.reflect.Modifier.STATIC)) Modifier.STATIC else 0)
    | (if (isOn(src, java.lang.reflect.Modifier.SYNCHRONIZED)) Modifier.SYNCHRONIZED else 0)
    | (if (isOn(src, java.lang.reflect.Modifier.ABSTRACT)) Modifier.ABSTRACT else 0)
    | (if (isOn(src, java.lang.reflect.Modifier.FINAL)) Modifier.FINAL else 0))
  }

  private def isOn(modifier: Int, flag: Int): Boolean =  (modifier & flag) != 0

  private final val CONSTRUCTOR_NAME: String = "<init>"
}

class ReflectionalClassTypeRef(klass: Class[_], table: ClassTable) extends IRT.AbstractClassType {
  import ReflectionalClassTypeRef._
  private val bridge: OnionTypeConversion              = new OnionTypeConversion(table)
  private val modifier_ : Int                          = toOnionModifier(klass.getModifiers)
  private var methods_ : MultiTable[IRT.Method]        = null
  private var fields_ : OrderedTable[IRT.FieldRef]     = null
  private var constructors_ : List[IRT.ConstructorRef] = null

  def isInterface: Boolean = (klass.getModifiers & java.lang.reflect.Modifier.INTERFACE) != 0

  def modifier: Int = modifier_

  def name: String = klass.getName

  def superClass: IRT.ClassType = {
    val superKlass: Class[_] = klass.getSuperclass
    if (superKlass == null) return table.rootClass
    val superClass: IRT.ClassType = table.load(superKlass.getName)
    if (superClass eq this) return null
    superClass
  }

  def interfaces: Array[IRT.ClassType] = {
    val interfaces: Array[Class[_]] = klass.getInterfaces
    val interfaceSyms: Array[IRT.ClassType] = new Array[IRT.ClassType](interfaces.length)
    for(i <- 0 until interfaces.length) {
      interfaceSyms(i) = table.load(interfaces(i).getName)
    }
    interfaceSyms
  }

  def methods: Array[IRT.Method] = {
    requireMethodTable
    methods_.values.toArray
  }

  def methods(name: String): Array[IRT.Method] = {
    requireMethodTable
    methods_.get(name).toArray
  }

  private def requireMethodTable {
    if (methods_ == null) {
      methods_ = new MultiTable[IRT.Method]
      for (method <- klass.getMethods) {
        if (!(method.getName == CONSTRUCTOR_NAME)) {
          methods_.add(translate(method))
        }
      }
    }
  }

  def fields: Array[IRT.FieldRef] = {
    requireFieldTable
    fields_.values.toArray
  }

  def field(name: String): IRT.FieldRef = {
    requireFieldTable
    fields_.get(name).getOrElse(null)
  }

  private def requireFieldTable {
    if (fields == null) {
      fields_ = new OrderedTable[IRT.FieldRef]
      for (field <- klass.getFields) {
        fields_.add(translate(field))
      }
    }
  }

  def constructors: Array[IRT.ConstructorRef] = {
    if (constructors_ == null) {
      constructors_ = new ArrayList[IRT.ConstructorRef]
      for (method <- klass.getConstructors) {
        constructors_.add(translate(method))
      }
    }
    constructors_.toArray(new Array[IRT.ConstructorRef](0))
  }

  private def translate(method: Method): IRT.Method = {
    val arguments: Array[Class[_]] = method.getParameterTypes
    val argumentRefs: Array[IRT.Type] = new Array[IRT.Type](arguments.length)
    for(i <- 0 until arguments.length) {
      argumentRefs(i) = bridge.toOnionType(arguments(i))
    }
    val returnRef: IRT.Type = bridge.toOnionType(method.getReturnType)
    new ClassFileMethodRef(toOnionModifier(method.getModifiers), this, method.getName, argumentRefs, returnRef)
  }

  private def translate(field: Field): IRT.FieldRef = {
    new ClassFileFieldRef(toOnionModifier(field.getModifiers), this, field.getName, bridge.toOnionType(field.getType))
  }

  private def translate(constructor: Constructor[_]): IRT.ConstructorRef = {
    val arguments: Array[Class[_]] = constructor.getParameterTypes
    val argumentRefs: Array[IRT.Type] = new Array[IRT.Type](arguments.length)
    for(i <- 0 until arguments.length) {
      argumentRefs(i) = bridge.toOnionType(arguments(i))
    }
    new ClassFileConstructorRef(toOnionModifier(constructor.getModifiers), this, "<init>", argumentRefs)
  }
}